### PR TITLE
Check for non nill attyid when pulling all judges

### PR DESCRIPTION
**Pull request by hschallhorn**
Resolves issue were non judge appears in a dropdown that should only contain judges (and acting judges). See investigation here: https://hackmd.io/@hschallh/By1eoWCO8
>Long story shorter:
> - When we pull a bulk list of judges from vacols (how we populate the dropdown), we only look at the user’s svlj  value of their vacols staff record.
> - When we check if someone is an acting judge in caseflow, we look at the svlj  AND sattyid of their vacols staff record.

From the board: All attorneys, judges, and acting judges should have a non nil sattyid

### Description
Create some convenience scopes and only return users with non nil sattyid when pulling all judges.

### Acceptance Criteria
- [ ] Tests pass!

### Testing Plan
1. Remove the sattyid from a judge's staff record
```ruby
user = FactoryBot.create(:user, css_id: "NOT_A_JUDGE", station_id: 101)
FactoryBot.create(:staff, svlj: "A", sattyid: nil, user: user)
```
2. Ensure they are not returned in Judge.list_all
```ruby
judges_for_dropdown = Judge.list_all
judges_for_dropdown.include? user
=> false
# Check for aaaaabshire just in case
judges_for_dropdown.include? User.find(3)
=> true
```
### User Facing Changes
- [ ] Non judges will no longer appear in judge only dropdowns

### CODE REVIEW

File: app/models/cached_user.rb [hunk 1]
```diff
@@ -12,7 +12,7 @@ class CachedUser < CaseflowRecord
 
   class << self
     def sync_from_vacols
-      VACOLS::Staff.where.not(sdomainid: nil).find_each do |staff|
+      VACOLS::Staff.having_css_id.find_each do |staff|
```
Comment (hschallhorn) [inline app/models/cached_user.rb:15]: Replace with new scope

File: app/models/vacols/staff.rb [hunk 2]
```diff
@@ -3,5 +3,19 @@
...
   scope :load_users_by_css_ids, ->(css_ids) { where(sdomainid: css_ids) }
+  scope :find_by_css_id,        ->(css_id)  { find_by(sdomainid: css_id) }
+  scope :active,                ->          { where(sactive: "A") }
+  scope :having_css_id,         ->          { where.not(sdomainid: nil) }
+  scope :having_attorney_id,    ->          { where.not(sattyid: nil) }
+  scope :pure_judge,            ->          { active.having_attorney_id.where(svlj: "J") }
+  scope :acting_judge,          ->          { active.having_attorney_id.where(svlj: "A") }
+  scope :pure_attorney,         ->          { active.having_attorney_id.where(svlj: nil) }
+  scope :judge,                 ->          { pure_judge.or(acting_judge) }
+  scope :attorney,              ->          { pure_attorney.or(acting_judge) }
```
Comment (hschallhorn) [inline app/models/vacols/staff.rb:16]: Scopes now clarify what user types will have what properties in their vacols records.

File: app/repositories/attorney_repository.rb [hunk 3]
```diff
@@ -2,22 +2,16 @@
...
-    end
-
     private
 
-    # returns CSS_IDs of attorneys and acting judges (who are normally attorneys)
+    # Returns CSS_IDs of pure attorneys and acting judges (who are normally attorneys)
     def find_all_having_attorney_ids_excluding_judges
-      VACOLS::Staff.where(sactive: "A").where.not(sdomainid: nil).where.not(sattyid: nil)
-        .where(svlj: [nil, "A"]).pluck(:sdomainid).map(&:upcase)
+      VACOLS::Staff.css_ids_from_records(VACOLS::Staff.attorney)
```
Comment (hschallhorn) [inline app/repositories/attorney_repository.rb:14]: `where.not(sdomainid: nil).pluck(:sdomainid).map(&:upcase)` -> `css_ids_from_records`
`where(sactive: "A").where.not(sattyid: nil).where(svlj: [nil, "A"])` -> `.attorney`

File: app/repositories/judge_repository.rb [hunk 4]
```diff
@@ -1,30 +1,22 @@
 # frozen_string_literal: true
 
 class JudgeRepository
-  # includes acting judges, who are normally attorneys
+  # Includes acting judges, who are normally attorneys
   def self.find_all_judges
-    css_ids = judge_records.where.not(sdomainid: nil).pluck(:sdomainid).map(&:upcase)
+    css_ids = VACOLS::Staff.css_ids_from_records(VACOLS::Staff.judge)
```
Comment (hschallhorn) [inline app/repositories/judge_repository.rb:6]: `where.not(sdomainid: nil).pluck(:sdomainid).map(&:upcase)` -> `css_ids_from_records`

File: app/repositories/judge_repository.rb [hunk 5]
```diff
@@ -1,30 +1,22 @@
...
         vacols_attorney_id: record.sattyid }
     end
   end
-
-  # returns judges and acting judges
-  # svlj="J" indicates judges
-  # svlj="A" indicates acting judges, who are normally attorneys
-  # svlj=nil indicates attorney
-  def self.judge_records
-    VACOLS::Staff.where(svlj: %w[J A], sactive: "A")
```
Comment (hschallhorn) [inline app/repositories/judge_repository.rb:28]: Source of the error. We were not checking for non nil attyids. `VACOLS::Staff.judge` now checks that.

File: app/repositories/user_repository.rb [hunk 6]
```diff
@@ -68,7 +68,7 @@ def css_id_by_full_name(full_name)
     def cached_staff_record(css_id)
       # Staff records rarely get updated so caching in Redis for 24 hours
       Rails.cache.fetch("#{Rails.env}_staff_record_#{css_id}") do
-        VACOLS::Staff.find_by(sdomainid: css_id)
+        VACOLS::Staff.find_by_css_id(css_id)
```
Comment (hschallhorn) [inline app/repositories/user_repository.rb:71]: Replace with new scope

File: spec/factories/task.rb [hunk 7]
```diff
@@ -384,14 +384,12 @@
 
         after(:build) do |_task, evaluator|
           if evaluator.assigned_by
-            existing_staff_record = VACOLS::Staff.where(
-              sdomainid: evaluator.assigned_by.css_id, svlj: "J", sactive: "A"
-            ).first
+            existing_staff_record = VACOLS::Staff.pure_judge.find_by_css_id(evaluator.assigned_by.css_id)
```
Comment (hschallhorn) [inline spec/factories/task.rb:387]: Replace with new scope

File: app/models/vacols/staff.rb [hunk 8]
```diff
@@ -3,5 +3,19 @@
...
+
+  def self.css_ids_from_records(staff_records)
+    staff_records.having_css_id.pluck(:sdomainid).map(&:upcase)
+  end
```
Comment (yoomlam) [inline app/models/vacols/staff.rb:20]: Very slight concern that method name may imply the returned count equals staff_records.count. In other words, method name doesn't suggest filtering may be happening.

Comment (hschallhorn) [reply app/models/vacols/staff.rb:20]: ++ Open to suggestions

File: app/repositories/attorney_repository.rb [hunk 9]
```diff
@@ -2,22 +2,16 @@
...
+    # Includes acting judges, who are normally attorneys
     def find_all_attorneys
       User.batch_find_by_css_id_or_create_with_default_station_id(find_all_having_attorney_ids_excluding_judges)
     end
 
-    # returns CSS_IDs of those with attorney_ids, including attorneys and judges
-    def find_all_having_attorney_ids
-      VACOLS::Staff.where(sactive: "A").where.not(sdomainid: nil).where.not(sattyid: nil)
-        .pluck(:sdomainid).map(&:upcase)
-    end
```
Comment (yoomlam) [inline app/repositories/attorney_repository.rb:13]: So all active attorneys, judges, and acting judges that we care about should have attorney_ids.

File: spec/repositories/judge_repository_spec.rb [hunk 10]
```diff
@@ -5,6 +5,7 @@
     judges = create_list(:staff, 3, :judge_role)
     acting_judges = create_list(:staff, 3, :attorney_judge_role)
     attorneys = create_list(:staff, 3, :attorney_role)
+    reporting_only_users = create_list(:staff, 3, :attorney_judge_role, sattyid: nil)
```
Comment (yoomlam) [inline spec/repositories/judge_repository_spec.rb:8]: Did you intend these users to be `:attorney_judge_role`? These users are acting judges with no attorney_id.
The tests still pass with that role removed: `create_list(:staff, 3)`.

**Comment by yoomlam** [comment]
My extra testing:
```ruby
j=FactoryBot.create(:user, :judge, css_id: "TEST_JUDGE1")
j.judge?
=> false  # unfortunate!

judge=FactoryBot.create(:user, css_id: "TEST_JUDGE2")
judge.judge?
=> false
sj=FactoryBot.create(:staff, :judge_role, user: judge)
judge.judge?
=> true

Judge.list_all.include? judge
=> false

# Why not?  Due to caching

JudgeRepository.find_all_judges.count
=> 225
Judge.list_all.count
=> 222

Rails.cache.clear
Judge.list_all.count
=> 225  # Better!

Judge.list_all.include? judge
=> true

judge.vacols_roles
=> ["judge"]

j.vacols_roles
=> []
Judge.list_all.include? j
=> false


aj = FactoryBot.create(:user, css_id: "ACTING_JUDGE-1")
FactoryBot.create(:staff, :attorney_judge_role, user: aj)
aj.judge?
=> true
aj.acting_judge_in_vacols?
=> true
aj.attorney_in_vacols?
=> true
aj.judge_in_vacols?
=> true

Rails.cache.clear
Judge.list_all.include? aj
=> true

```

File: spec/repositories/judge_repository_spec.rb (hunk 10 above)
Comment (hschallhorn) [reply spec/repositories/judge_repository_spec.rb:8]: Yes! If you look at the investigation, the bug that brought this up is that there are users in vacols with an svlj of "A" but no attorney id. There are over 60 of these users. This test would have failed without this PR. With your updated `create_list(:staff, 3)` these tests would have passed with our old code. This is here to ensure staff with a svlj of "A" are not returned in Judge.list_all

**Comment by hschallhorn** [comment]
@yoomlam Am I missing the changes that need to be made?

File: spec/repositories/judge_repository_spec.rb (hunk 10 above)
Comment (yoomlam) [reply spec/repositories/judge_repository_spec.rb:8]: Gotcha!  I recommend renaming `reporting_only_users` to `acting_judges_without_attyid`. I don't know how to interpret `reporting_only_users`.

**Pull request merged**